### ISSUE TITLE
arbitrator challenge logic (audit #6)

### ISSUE THREAD

**Issue by jellegerbrandy** [hats-finance/hats-contracts#233]
> Adam Kolar | g0, [09/06/2022 17:57]
> 6) So I don't think the conditions in approveClaim and dismissClaim are quite right. Because:
> - claim can be approved by arbitrator even before challengePeriod expires, but only if it's challenged, so if arbitrator wants to approve claim before challenge period expires, they have to challenge it first. This seems strange.
> - after challengePeriod expires it can be approved by anybody even if it has been challenged, this might be intentional. But I'd expect challenged claims to be only approvable by arbitrator or after challengeTimeOutPeriod expires. With current functionality a dismissalPeriod would be more apt name than challengePeriod.
> - claim can be dismissed by anybody after challengeTimeOutPeriod expires even if it hasn't been challenged. Again this might be intentional.
> - after challengePeriod expires, it can be approved by anybody, but it can also be challenged and dismissed by arbitrator.
> 
> Adam Kolar | g0, [09/06/2022 18:02]
> To put it in different terms, whether or not a claim is challenged currently affects only whether a claim can be dismissed and approved by arbitrator. Otherwise claim being challenged has no effect on the claim.
> 
> Adam Kolar | g0, [09/06/2022 18:04]
> With this in mind and if this functionality is intentional, challengeTimeOutPeriod should be renamed to claimTimeOutPeriod and challengePeriod should be renamed dismissalPeriod.

**Comment by jellegerbrandy** [comment]
So the context here is that `arbitrator` is a contract or set of contracts, such as Kleros or reality.eth

* "claim can be approved by arbitrator even before challengePeriod expires, but only if it's challenged, " -> this is ok. 
* "after challengePeriod expires it can be approved by anybody even if it has been challenged, .." **this is a real bug**

* "claim can be dismissed by anybody after challengeTimeOutPeriod expires" -> this is as intended
* "after challengePeriod expires, it can be approved by anybody, but it can also be challenged and dismissed by arbitrator" -> yes, this is not necessarily wrong I think 
This line in `approveClaim` needs to change:
```
  if (!(msg.sender == arbitrator && claim.isChallenged) &&
            (claim.createdAt + challengePeriod >= block.timestamp))
        revert ClaimCanOnlyBeApprovedAfterChallengePeriodOrByArbitrator();

 ```
And can become:
```
  if (msg.sender != arbitrator && (claim.isChallenged || claim.createdAt + challengePeriod >= block.timestamp)) revert ClaimCanOnlyBeApprovedAfterChallengePeriodOrByArbitrator();
```

**Comment by ben-kaufman** [comment]
With the condition as you now suggest it, the arbitrator will be able to approve even without the claim being challenged, is that intentional?

**Comment by ben-kaufman** [comment]
Also on the last point, I think this is bad as it creates some partly sort of a race condition, but maybe not really a problem in this context

**Comment by jellegerbrandy** [comment]
> With the condition as you now suggest it, the arbitrator will be able to approve even without the claim being challenged, is that intentional?

Yes it was (given that arbitrator is also on charge of challenging it seems silly to rquire the arb to challenge first) but now I think of it is is cleaner to make it really rexplicit as specced:

```
if claim.isChallenged then {
   if msg.sender != arbitrator then revert
} else {
   if block.timestamp <= claim.createdAt + challengePeriod then revert;
}
```
